core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -1865,6 +1871,62 @@ abstract class RDD[T: ClassTag](
   // RDD chain.
   @transient protected lazy val isBarrier_ : Boolean =
     dependencies.filter(!_.isInstanceOf[ShuffleDependency[_, _, _]]).exists(_.rdd.isBarrier())
+
+  /**
+   * Returns the random level of this RDD's output. Please refer to [[RandomLevel]] for the
+   * definition.
+   *
+   * By default, an reliably checkpointed RDD, or RDD without parents(root RDD) is IDEMPOTENT. For
+   * RDDs with parents, we will generate a random level candidate per parent according to the
+   * dependency. The random level of the current RDD is the random level candidate that is random
+   * most. Please override [[getOutputRandomLevel]] to provide custom logic of calculating output
+   * random level.
+   */
+  // TODO: make it public so users can set random level to their custom RDDs.
+  // TODO: this can be per-partition. e.g. UnionRDD can have different random level for different
+  // partitions.
+  private[spark] final lazy val outputRandomLevel: RandomLevel.Value = {
+    if (checkpointData.exists(_.isInstanceOf[ReliableRDDCheckpointData[_]])) {
```

Comment:
hmm, so I took another look at the checkpoint code, and it seems to me like it doesn't checkpointing will actually help.  IIUC, checkpointing doesn't actually take place until the *job* finishes, not just the stage:

https://github.com/apache/spark/blob/6193a202aab0271b4532ee4b740318290f2c44a1/core/src/main/scala/org/apache/spark/SparkContext.scala#L2061-L2063

So when you have a failure in the middle of a job with a long pipeline, when you go back to an earlier stage, you're not actually going back to checkpointed data.

But maybe I'm reading this wrong?  doesn't seem like what checkpointing _should_ be doing, actually ...

Reply:
yeah though I guess I'm also saying, this problem is worse than we expected, as checkpointing is not a good way to cope with the added cost.

Reply:
yes, as far as I remember its like caching, you have to force an action to get it to checkpoint early.  So its still possible its just you have to be more explicit.